kore/src/Kore/Internal/SideCondition.hs

```diff
@@ -577,3 +620,234 @@ retractLocalFunction =
             InternalSet_ _ -> Just (Pair term1 term2)
             _          -> Nothing
     go _ _ = Nothing
+
+{- | Assumes a 'TermLike' to be defined. If not always defined,
+it will be stored in the `SideCondition` together with any subterms
+resulting from the implication that the original term is defined.
+
+For maps and sets: this will generate and store a minimal set
+of sub-collections from which the definedness of any sub-collection
+can be inferred.
+-}
+assumeDefined
+    :: forall variable
+    .  InternalVariable variable
+    => TermLike variable
+    -> SideCondition variable
+assumeDefined term =
+    assumeDefinedWorker term
+    & fromDefinedTerms
+  where
+    assumeDefinedWorker
+        :: TermLike variable
+        -> HashSet (TermLike variable)
+    assumeDefinedWorker term' =
+        case term' of
+            TermLike.And_ _ child1 child2 ->
+                let result1 = assumeDefinedWorker child1
+                    result2 = assumeDefinedWorker child2
+                 in asSet term' <> result1 <> result2
+            TermLike.App_ symbol children ->
+                checkFunctional symbol term'
+                <> foldMap assumeDefinedWorker children
+            TermLike.Ceil_ _ _ child ->
+                asSet term' <> assumeDefinedWorker child
+            TermLike.InternalList_ internalList ->
+                asSet term'
+                <> foldMap assumeDefinedWorker (internalListChild internalList)
+            TermLike.InternalMap_ internalMap ->
+                let definedElems =
+                        getDefinedElementsOfAc internalMap
+                    definedMaps =
+                        generateNormalizedAcs internalMap
+                        & HashSet.map TermLike.mkInternalMap
+                 in foldMap assumeDefinedWorker definedElems
+                    <> asSet term'
+                    <> definedMaps
+            TermLike.InternalSet_ internalSet ->
```

Comment:
Could the logic for Map and Set be combined with `Kore.Builtin.AssocComm.CeilSimplifier`?

Reply:
I've thought about this, and I think it would only make sense if we wanted to also generate subcollections there and evaluate their ceils. Do we want to do this? At the moment, it looks like we only check the definedness of the elements and if they are unique.